source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,384 @@
+
+    public this (ref Transaction[] txs)
+    {
+        MerkleNode[] nodes;
+
+        foreach (ref tx; txs)
+            nodes ~= MerkleNode(null, null, &tx);
+
+        const max_level = to!long(ceil(log2(nodes.length)));
+        for (int i = 0; i < max_level; i++)
+        {
+            if (nodes.length % 2 != 0)
+                nodes ~= nodes[$ - 1];
+
+            MerkleNode[] nodesInLevel;
+            for (int j = 0; j < nodes.length; j += 2)
+                nodesInLevel ~= MerkleNode(&nodes[j], &nodes[j + 1]);
+            nodes = nodesInLevel[];
+        }
+
+        if (nodes.length > 0)
+            this.root = &nodes[0];
+        else
+            this.root = null;
+    }
+
+    /*******************************************************************************
+
+        Found nodes with specific hash on child nodes.
+
+        Params:
+            hash = looking hash.
+
+        Returns:
+            Return node if found the hash. Return null otherwise.
+
+    *******************************************************************************/
+
+    public MerkleNode* searchNode (Hash hash)
+    {
+        return this.root.searchNode(hash);
+    }
+
+
+    /*******************************************************************************
+
+        Get Merkle path
+
+        Params:
+            hash = `Hash` of `Transaction`
+
+        Returns:
+            Return array of `Hash` of Merkle path
+
+    *******************************************************************************/
+
+    public Hash[] getMerklePath (Hash hash)
+    {
+        bool getPath (MerkleNode* node, Hash hash, ref Hash [] path)
+        {
+            if (node is null) return false;
+
+            if (node.hash == hash) return true;
+
+            if (getPath(node.right, hash, path))
+            {
+                path ~= node.left.hash;
+                return true;
+            }
+
+            if (getPath(node.left, hash, path))
+            {
+                path ~= node.right.hash;
+                return true;
+            }
+
+            return false;
+        }
+
+        Hash [] path;
+        getPath(this.root, hash, path);
+        return path;
+    }
+
+    /*******************************************************************************
+
+        Prove merkle path
+
+        Params:
+            hash = `Hash` of `Transaction`
+            path = `Hash` of Merkle path
+            root = `Hash` of Merkle root
+
+        Returns:
+            Return true if the Merkle proof is completed. Return false otherwise.
+
+    *******************************************************************************/
+
+    static public bool proveMerklePath (Hash hash, Hash[] path, Hash root, int index)
+    {
+        if (path.length == 0) return false;
+        if (index < 0) return false;
+
+        Hash h = hash;
+        int idx = index;
+        foreach (ref p; path)
+        {
+            if (idx % 2 == 0)
```

Comment:
I'm not sure I understand the need for `index` here ?

Reply:
To distinguish between parameters and internal variables.

Reply:
OK